src/west/app/project.py

```diff
@@ -867,6 +867,15 @@ def fetch_missing_imports(self, args):
         self.manifest = Manifest.from_file(topdir=self.topdir,
                                            importer=self.update_importer)
 
+    def update_submodules(self, project, submodules, call_location):
+        # Updates given list of project submodules by using
+        # 'git submodule update --init' command from the location passed
+        # through the call_location parameter value.
+        if submodules:
+            for submodule in submodules:
+                project.git(['-C', call_location, 'submodule', 'update',
```

Comment:
A few questions here:

- Don't you need a `--checkout` here to make sure the submodule isn't rebased if `submodule.<name>.update` is `rebase`, for example?
- What about recursion into submodules of submodules? Should this be optional, somehow?
- How can we ensure the submodule is checked out as a detached HEAD?

Reply:
Honestly I don't think I'm able to answer it, as I wasn't assuming that it is needed. As far as I understand, you suggest that any local changes done in submodules should be overridden by the update and current code version will not do this. I'm really not very into detailed west operation to judge whether it's good or not, so I believe that I should rely on your statement. Could you share your vision about this functionality to let me understand it better? If it comes to the recursion - I think that could be useful think, but do you think that it should be selected by setting some value in the manifest or e.g. by calling west update command with some additional argument (-r is already taken by the rebase option).

Reply:
> As far as I understand, you suggest that any local changes done in submodules should be overridden by the update and current code version will not do this. I'm really not very into detailed west operation to judge whether it's good or not,

I think consistency between submodules and west "modules" is very important and should always be the default unless the user very explicitly and clearly requests `west` and `submodules` to behave differently. Whatever `west update` does, `git submodule` should do the  same otherwise it will cause a lot of confusion.

(Just FYI I almost always use `west update -r`, I find `west update` without `-r` crazy :-)

> What about recursion into submodules of submodules? Should this be optional, somehow?

Maybe optional (then good luck for extensive test coverage) but again it should have the exact same **default** behaviour as `west`. Imagine a complex project with more than one level of `west` manifests and then multiple levels of git submodules somewhere underneath. Now imagine that `west update` imports all the manifests by default but for some strange reason updates only one level of submodules by default. That line would feel arbitrary and be quite confusing.

Reply:
@marc-hb thanks, now I see potential problems more clearly. Then I assume that probably better option would be to make recursion on by default, to make sure that none of the levels was skipped, and allow disabling it in manifest, right?

Reply:
> Then I assume that probably better option would be to make recursion on by default,

I think yes, recursive by default. However I've used west imports only once for pure testing purposes so far. So for that and other reasons please wait for @mbolivar-nordic

> and allow disabling it in manifest, right?

Depends what you mean by "disabling". Disabling this entire new submodules feature completely for fine-grained user control and reverting to older behaviours sounds nice to me. Not sure where it should be done: in the top level/other manifest, user config file, other, all of the above with well defined precedence...? The devil is in the details.

Disabling recursive updates which would leave only the first level of submodule seems like a confusing and bad setting to me because as I wrote before this would make the feature "recursive until the first levels of submodules". 

Here's how a "normal" recursive option works:
- yes: works on all levels.
- no: works on 1 level

Here's how a weird, recursive submodule setting would work in west:
- yes: works on all west imports + all submodule levels
- no: works on all west imports + 1 submodule level ?!?

I don't think anyone would enjoy this feature.

Reply:
Hi,

> As far as I understand, you suggest that any local changes done in submodules should be overridden by the update and current code version will not do this. 

Not exactly. By default, west uses `git checkout --detach` to update each project to the new manifest revision. This does not affect any uncommitted working tree changes as long as they don't conflict with the checked out revision.

It does leave behind any commits that aren't part of the new manifest-rev unless you use `-r`. It leaves a detached HEAD unless you use `-r` or sometimes when you use `-k`.

I agree with @marc-hb that "west with submodules" should try to at least keep the default behavior:

1. leave a detached HEAD checked out
2. leave any uncommitted working tree changes alone
3. error out if uncommited changes conflict with the new submodule revision

Is that possible to do with git submodules?

> I think yes, recursive by default. However I've used west imports only once for pure testing purposes so far. So for that and other reasons please wait for @mbolivar-nordic

I think recursing into submodules of submodules is a good idea if it is possible / practical to get "everything" by default.

We can always add additional configuration to do more complicated stuff:

```yaml
manifest:
  projects:
  - name: foo
    submodules:
      recurse-behavior: whatever
```

Note I am **not** suggesting we add a `recurse-behavior` key right now. I'm just saying we could do it later if we need finer grained control.

But I am not a submodules user (hey, I just use west...) so I don't know if that addresses people's use cases and I want the users to chime in here. Thanks.

Reply:
@marc-hb @mbolivar-nordic great tips, I really appreciate it. I added `--recursive` option as a default one, because for now it seems to be the best option and I also added test case for that in one of tests.

> It does leave behind any commits that aren't part of the new manifest-rev unless you use -r. It leaves a detached HEAD unless you use -r or sometimes when you use -k.

I agree with @marc-hb that "west with submodules" should try to at least keep the default behavior:

    leave a detached HEAD checked out
    leave any uncommitted working tree changes alone
    error out if uncommited changes conflict with the new submodule revision

Is that possible to do with git submodules?
>

Actually current behavior of submodules with `--checkout` option is almost the same, as described. I wrote almost, because it leaves at detached HEAD state, leaves uncommitted changes alone, prints error in case of conflicts and overrides new commits which are not part of new manifest-rev, but it doesn't care about `-r` or `-k` option. I tried to fix it using `--rebase` and `--checkout` options, but few problems occurred:

- Current west update strategy assumes that we should make rebase or keep descendants only when project is not in the detached state. That implies problem when some changes are done in the project's submodule and it should be rebased, but project itself is not modified, so rebase will not be selected. That's not a big problem at all, because new method introducing exact same behavior for submodules could be implemented to solve it.

- Real problem is using `--recursive` option, because it implies that all submodules of submodule will be updated the same way, as submodule is (with `--rebase`, `--checkout` or anything else). That means that when using `-r` submodules on the first level will behave same way as projects (so rebase when modified or checkout otherwise) and on other levels will always do the same as parent submodule did, without possibility to select other strategy. This will raise confusion mentioned by @marc-hb.  

What are your opinions on that? Do you think making assumption that submodules are always checked out and don't support rebase is acceptable? Summing up - this way default `west update` should work properly, but command extensions unfortunately not.

Reply:
> What are your opinions on that? Do you think making assumption that submodules are always checked out and don't support rebase is acceptable?

I'll try to summarize to check my understanding: I think this means the west project update respects `-k` and `-r`, but submodules are always checked out as detached HEADs to their "gitlink" revisions. Submodule update behavior is not affected by `-k` or `-r`.

I find this acceptable: it's consistent, and treats west projects as first class citizens, with submodules just being additional "bags on the side" that follow the west project strictly, which is easy to understand and explain.

@marc-hb ?

Reply:
Pinging @sylvioalves on this question also, since @nashif brought him in. I think this comment thread is the last major unresolved question.

Reply:
> I think this means the west project update respects -k and -r, but submodules are always checked out as detached HEADs to their "gitlink" revisions. Submodule update behavior is not affected by -k or -r.

That's not what I understood. What I understood in a nutshell  (I could be wrong): the behavior of `west --rebase` and `west --keep-descendants` is project-specific (depending on whether there is a branch) whereas `submodule --rebase` is unconditional and this can creates funny inconsistencies.

> treats west projects as first class citizens, with submodules just being additional "bags on the side" that follow the west project strictly, which is easy to understand and explain.

I'm sure there will be `west/submodule` inconsistencies unfortunately impossible to address because these are two different tools with two different designs. There should just be as few inconsistencies as possible. I'm afraid `--rebase` and `--keep-descendants` might indeed end in that minimal list of unavoidable inconsistencies.  In such cases I think it's best to keep at least all submodules behaving the same, no special treatment for  the first submodule level for instance. Warnings are also useful for such complex situations and they save code comments.

I think the documentation of important inconsistencies like this potential one should be documented/specified in this PR already. I find that writing documentation is a great way to take some distance from the implementation details and make sure the high-level design is sound _and_ user friendly.

Reply:
> I'll try to summarize to check my understanding: I think this means the west project update respects -k and -r, but submodules are always checked out as detached HEADs to their "gitlink" revisions. Submodule update behavior is not affected by -k or -r.
>

@mbolivar-nordic Yes, you are absolutely right. I will try to clarify it even more, to dispel any doubts. Currently for **every** project `west update` simplified behavior looks like this:

`
if current_branch != HEAD:
    if k_option_enabled:
        do_nothing()
    else if r_option_enabled:
        rebase()
    else:
        checkout()
else:
    checkout()
`

Update method cares about `-k` and `-r` options only when current branch is not in the HEAD state and otherwise they are ignored.

> That's not what I understood. What I understood in a nutshell (I could be wrong): the behavior of west --rebase and west --keep-descendants is project-specific (depending on whether there is a branch) 
>
That's true.

> whereas submodule --rebase is unconditional and this can creates funny inconsistencies.
>
@marc-hb I'm not 100% sure what you meant, but submodule --rebase currently is not used and submodules **always** do checkout, no matter of what parent project does.

I could make submodules to behave the same as projects and respect `-r` and `-k` options, but only at first level, because of the `--recursive` option, but I think that would be wrong, as you wrote:

>  In such cases I think it's best to keep at least all submodules behaving the same, no special treatment for the first submodule level for instance.
>

If it comes to the documentation - of course I can add it, but @mbolivar-nordic declared that he would like to do it on his own.

Reply:
> I could make submodules to behave the same as projects and respect -r and -k options, but only at first level, because of the --recursive option

I just performed a quick test and `git submodule update --rebase --recursive` does rebase recursively:
```
git submodule update --rebase --recursive
First, rewinding head to replay your work on top of it...
Applying: mycommit
Submodule path 'rimage': rebased into 'fccf52133104987d9a532e6a9c332ba5c7a1bd91'
First, rewinding head to replay your work on top of it...
Applying: mycommit
Submodule path 'rimage/tomlc99': rebased into 'e3a03f5ec7d8d33be705c5ce8a632d998ce9b4d1'
```
Note there was a branch in both repos; not sure it matters for submodules.

According to `git help submodule`, `--rebase` and `--checkout` seem mutually exclusive.

> If it comes to the documentation - of course I can add it, but @mbolivar-nordic declared that he would like to do it on his own.

I certainly don't want to get into "who should do what" while I'm only offering incomplete reviews from a safe distance but I think the high-level logic should be somewhat documented or in other words "specified" before this gets merged because I don't think it's possible to review the code without the complete knowledge of how it's supposed to behave. That documentation could be just a draft, maybe not in this PR then. I mean something looking exactly like what you just did here:

> if k_option_enabled:
>        do_nothing()

This particular bit seems a bit too simplified :-)

Reply:
> This means that when using -r submodules on the first level will behave same way as projects (so rebase when modified or checkout otherwise) and on other levels will always do the same as parent submodule did, without possibility to select other strategy. 

OK, I think I finally understood this. I got confused by "only on the first submodule level". 

I think what you're saying is: `west --rebase` can be repo-specific while `submodule --rebase --recursive` always performs a rebase, it's never repo-specific. That's one unavoidable inconsistency between `west` and `submodule`.

Considering `submodule --recursive` always performs the same action when traversing down, I think we all agree that all `submodule` invocations from west should all be given the same parameters, not depending on decisions `west` made in west repos. This keeps all submodules behaving the same no matter what.

However I think I disagree on never passing `--rebase` to `submodule`. I think it's more intuitive (and IMHO safer) to always pass `--rebase` through `west` and to `submodule` when and only when the user ask `west` to `--rebase`[*] 

Either way, the existence of some unavoidable inconsistency probably means  a "don't touch submodule" flag is needed sooner than later.

[*] I forgot why `west update --rebase` does not rebase detached HEADs. Feature or limitation?

Reply:
I'm not west specialist and I build my understanding of project policy on your comments and reading code, but what I understood is that the worst thing is making inconsistent features and in my opinion we are just starting to go this way.
> I think what you're saying is: west --rebase can be repo-specific while submodule --rebase --recursive always performs a rebase, it's never repo-specific. That's one unavoidable inconsistency between west and submodule.

Considering submodule --recursive always performs the same action when traversing down, I think we all agree that all submodule invocations from west should all be given the same parameters, not depending on decisions west made in west repos. This keeps all submodules behaving the same no matter what.
> 
I agree with this, making submodules update strategy dependent only on first level of submodules and forcing another levels to behave way as first does is not good idea and we should keeps all submodules behaving the same no matter what.

> However I think I disagree on never passing --rebase to submodule. I think it's more intuitive (and IMHO safer) to always pass --rebase through west and to submodule when and only when the user ask west to --rebase.
>
But that one is for me the field to make another inconsistency. We have two options (at least I see two):

- Do, as you suggest, what is for sure fine, because we will support `-r`option for submodules, but we can't support `-k` option, as I think we won't be able to check ancestors of all submodules passed recursively and make decision whether we should do the checkout or not. That raises another inconsistency between projects and submodules update strategy.

- Do not support `-r` and `-k` options for submodules, what means that they will behave like less advanced version of update for projects, which supports only default behavior resulting by invoking `west update` command without any additional arguments. In my opinion this one has got less functionality, but is more clearer.

> [*] I forgot why west update --rebase does not rebase detached HEADs. Feature or limitation?
>
I don't know, but it looks like in the code someone made an assumption that only when branch is not in the detached HEAD state, any other options than default checkout might be considered.

Reply:
I think it's intuitive and fairly consistent to pass-through a `--rebase` feature that has the exact same name and almost same behavior while dropping the `--keep-descendants` feature for the simple and intuitive reason that the latter does not exist in submodules. AFAIK the `--rebase` behavior difference is only for "careless" people who forget to save their precious work in a branch, correct? All other, careful users will not see any `--rebase` difference?

But more importantly: `--rebase` is not any random feature, it is a critical feature, because it's the only `--do-not-delete-all-my-work` feature available in both `west` and `submodule`. Consistency considerations aside, restricting it to `west` would be a great loss and block everyone from doing any work in submodules. Users who do not work in submodules and only use them as "bags on the side" don't care about `--rebase` etc. because they have nothing to rebase and won't be affected whatsoever.

In fact `--rebase` should have been a safer default in both. I digress.

Reply:
> > 
> 
> If it comes to the documentation - of course I can add it, but @mbolivar-nordic declared that he would like to do it on his own.

@ kkasperczyk-no  Feel free to send a patch to the zephyr repository with documentation for this feature! I would certainly appreciate it.

I was just trying to make less work for you, since we can't merge that patch until we're ready to release the next version of west. That is something I have to do, because it also involves rolling up API changes, writing a changelog, etc., so I was going to be doing docs anyway.

If you write a docs patch, I'll just cherry-pick it into the other docs work I will do for the next release, with gratitude :smile:.

However, I think simplified pseudocode descriptions of the feature's behavior are good enough for getting the PR reviewed and merged, with more polished docs to be written later.

> [*] I forgot why west update --rebase does not rebase detached HEADs. Feature or limitation?

@marc-hb Limitation, I suppose.

> Do not support -r and -k options for submodules, what means that they will behave like less advanced version of update for projects, which supports only default behavior resulting by invoking west update command without any additional arguments. In my opinion this one has got less functionality, but is more clearer.

I agree.

I propose a compromise: can we add a separate `--rebase-submodules` option that rebases the submodules if given, and otherwise we just continue to use `--checkout`?  This makes it clear that `-k` and `-r` are for west projects, while submodules have different behavior with their own set of options.

Reply:
> However, I think simplified pseudocode descriptions of the feature's behavior are good enough for getting the PR reviewed and merged, with more polished docs to be written later.

+1

> I propose a compromise: can we add a separate --rebase-submodules option that rebases the submodules if given, and otherwise we just continue to use --checkout? This makes it clear that -k and -r are for west projects, while submodules have different behavior with their own set of options.

I think this is an extra complication and hassle to highlight a rebase difference/west limitation that is not visible in "normal" usage. But it's certainly much better than not having the feature at all and being forced to drop all local commits in submodules.

If it's not the default then it will lose the users' commits but only the first times: after having tediously restored branches one by one or worse: inspected hard-to-read reflogs a few times to recover their work, the pain experienced should help them remember about the existence of `--rebase-submodules` for the next update and teach them the new lesson to always use branches even with submodules (in case they forget about the `--rebase-submodules` option again). Or they can configure `.gitmodules` but that's much less flexible: sometimes you do want to revert to some exact, untainted version.

> which supports only default behavior resulting by invoking west update command without any additional arguments. In my opinion this one has got less functionality, but is more clearer.

I understand that "-k and -r are not passed" is shorter and clearer than "-r is passed and -k is not" but that's a very small difference. It's not like the latter is difficult to remember considering submodules only support the latter (better than west).

In `git help submodule` you can see that `--rebase` is not considered like "an extra feature", it's shown at the exact same level than `--checkout` and `--rebase`. `--checkout` is just the default value, for backward compatibility I guess.

Reply:
> But more importantly: --rebase is not any random feature, it is a critical feature, because it's the only --do-not-delete-all-my-work feature available in both west and submodule. 
>
I think it convinces me. All in all saving our work from losing is one of the elementary repository use case and I might be more important than some structural assumptions.

> I understand that "-k and -r are not passed" is shorter and clearer than "-r is passed and -k is not" but that's a very small difference.
>
To be honest I don't like it in 100%, but I must admit that considering all aspects, which we discussed, each has got some disadvantages and following @marc-hb suggestion is probably the best one. It prevents user from losing work, has some inconsistency, but still is quite simple and doesn't require adding new commands not known for users (that for sure would result in wasting some users time on learning it). Anyway I think that we should decide something together to bring it to the end, so after reflection I'm ok with @marc-hb version. I will try to implement, test and introduce it, but if anybody is not ok with that, or we should try another approach, because of some reason, please share your opinion.

Reply:
Rebasing option and test for both checkout and rebase update strategies were added.

Reply:
> I'm ok with @marc-hb version. I will try to implement, test and introduce it, but if anybody is not ok with that, or we should try another approach, because of some reason, please share your opinion.

OK, I can see the arguments on either side and I don't have a strong opinion here, so I want the users' voices to carry the largest weight. I am fine with this.

@marc-hb if you have a chance to look at this and give it a +1 I'd appreciate it, same for @sylvioalves.

Reply:
Sure, let me review the solution approach and test locally..

Reply:
For the purpose I was looking for, it works.

Reply:
Just submitted a longer description of `--keep-descendants` in https://github.com/zephyrproject-rtos/zephyr/pull/31201

Reply:
> > I think what you're saying is: west --rebase can be repo-specific while submodule --rebase --recursive always performs a rebase, it's never repo-specific. That's one unavoidable inconsistency between west and submodule.
> 
> > Considering submodule --recursive always performs the same action when traversing down, I think we all agree that all submodule invocations from west should all be given the same parameters, not depending on decisions west made in west repos. This keeps all submodules behaving the same no matter what.
> 
> I agree with this, making submodules update strategy dependent only on first level of submodules and forcing another levels to behave way as first does is not good idea and we should keep all submodules behaving the same no matter what.

Jumping in here.

Why not use the existing infrastructure of `west` ?

Using `git submodule status` here, instead of `git submodule update --init --recursive` means we can internally create `west` projects.

Git stores the remote in `.gitmodules`, as example here:
https://github.com/ThrowTheSwitch/CMock/blob/master/.gitmodules#L7
```
[submodule "vendor/c_exception"]
	path = vendor/c_exception
	url = https://github.com/throwtheswitch/cexception.git
	branch = master
```
and a `git submodule status` will give you the SHA to checkout.
```
/.../cmock $ git submodule status
-dce9e8b26f2179439002e02d691429e81a32b6c0 vendor/c_exception
-031f3bbe45f8adf504ca3d13e6f093869920b091 vendor/unity
```
so you should be able to create those two projects a regular `west` projects, in which case the `-k` and `-r` will work.
Once the first level of submodules have been checked out and detached, you then enter the folder if `west` manifest has recursive == true.

As example:
```
# This would be done by `west` if the submodule is added as a project:
/.../cmock$ git -C vendor/c_exception/ init 
/.../cmock$ git -C vendor/c_exception/ fetch https://github.com/throwtheswitch/cexception.git master
 # SHA from above `git submodule status` command
/.../cmock$ git -C vendor/c_exception/ checkout dce9e8b26f2179439002e02d691429e81a32b6c0

# which means we can just do the same again, until no more submodules.
/.../cmock$ cd vendor/c_exception/
/.../cmock/vendor/c_exception$ git status
HEAD detached at dce9e8b
nothing to commit, working tree clean
/.../cmock/vendor/c_exception$  git submodule status
-2c7629a0ae90ffe991b5fd08e4db8672f72ed64c vendor/unity

/.../cmock/vendor/c_exception$  cat .gitmodules 
[submodule "vendor/unity"]
        path = vendor/unity
        url = https://github.com/throwtheswitch/unity.git
```

The `.gitmodules` file can easily be read using `configparser`, and creating `west` projects for each recursive submodule allows `west` to use exactly the same strategy on submodules as it does on any other `west` project.

I ackknowledge, this is quite a change to the current approach, but it would ensure `west` behaves identical bewteen `west` projects and submodules regarding `-k` and `-r`.